src/vm/methodtable.cpp

```diff
@@ -1594,6 +1592,13 @@ BOOL MethodTable::CanCastToClass(MethodTable *pTargetMT, TypeHandlePairList *pVi
     }
     CONTRACTL_END
 
+    // allow an object of type T to be cast to Nullable<T> (they have the same representation)
```

Comment:
It is not ok to add this here. It is changing how the type system works in subtle ways and it would require a thorough review and fixes/refactorings that I do not think you want to do as part of the change.

For example, the following is expected to throw exception, but it will succeed with your changes because of this:

```
using System;

class G<T,U> where T:U
{
}

class Test
{
    static void Main()
    {
        typeof(G<,>).MakeGenericType(typeof(int), typeof(int?));
    }
}
```

(Please add it as a test case while you are on it.)

Reply:
I thought I fixed this one. Even added a test. I think I misunderstood the issue. It is not about boxed/unboxed forms. It is about casting an object or casting a type.
Yeah, we need a test for this.

---
In reply to: [335189131](https://github.com/dotnet/coreclr/pull/23548#discussion_r335189131) [](ancestors = 335189131)